rfc/system/5818-decentralized-agent-config-mgmt.md

```diff
@@ -0,0 +1,147 @@
+
+
+```yaml
+apiVersion: v1
+king: ConfigMap
+name: flyteplugins-agentservice-config-xxx
+data:
+    plugins:
+      agent-service:
+        agents:
+          custom-agent:
+           endpoint: custom_agent_end_point_FQDN:8000
+           insecure: true
+        supportedTaskTypes:
+        - custom_sensor
+        - customtask
+```
+
+```yaml
+apiVersion: v1
+king: ConfigMap
+name: flyteplugins-agentservice-enabled-plugins-xxx
+data:
+    plugins:
+      agent-service:
+        supportedTaskTypes:
+        - customsensor
+        - customtask
+    tasks:
+      task-plugins:
+        default-for-task-types:
+          customtask: agent-service
+          customsensor: agent-service
+```
+
+### Aggregation and Conflict Resolution
+After validation, valid ConfigMaps should be aggregated. The system must handle potential conflicts (e.g., two agents with the same name or endpoint) by either resolving them or flagging them for review. This ensures that only conflict-free configurations are applied, isolating issues and preventing one bad agent configuration from affecting the entire Flyte setup.
+
+## 4. Metrics & Dashboards
+
+There can be different metrics to track the custom agent configuration validation such as total request, validated cases and its latency. These [metrics](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/webhook/internal/metrics/metrics.go) are natively supported in controller framework such as controller-runtime
+
+Any failures in aggregation and pre-checks of the agent reachability should be properly logged. K8s events can be emitted as well. This enables the enterprise environment to aggregate the logs/events for troubleshooting. 
+
+
+## 5. Drawbacks
+
+Are there reasons not to proceed with this proposal? 
+- What are the risks involved?
+
+* This requires the rewrite of viper in statically parsing the configuration files
+
+- Could decentralized management introduce more complexity than it solves?
+
+Instead of statically parse from the config files, the change is more or less to dynamically interact the K8s API to constructs the `Config` object
+
+
+
+## 6. Alternatives
+
+An alternative approach could be to specify the ConfigMap as a volume and mount the custom agents' ConfigMap in the Propeller pod. This way, the existing Viper parsing might still work, along with additional aggregation and merging capabilities.
+
+```yaml
+spec:
+  containers:
+  - name: propeller-container
+    image: image
+    volumeMounts:
+    - name: config-volume
+      mountPath: /etc/config/flyte/custom-agents/custom-agent-configmap.yaml
+  volumes:
+  - name: config-volume
+    configMap:
+      name: custom-agent-configmap
+```
+
+However, this still requires modifications to the centralized Propeller `Deployment` YAML file, which is similar to the current approach of directly updating the centralized Propeller ConfigMap. Thus, the same challenges will persist.
```

Comment:
But it's a one-time change, right? you just need to change the deployment yaml at the first time, and your team can own the `custom-agent-configmap`

Reply:
I would start with this approach since it’s much simpler and easier to implement. Additionally, it allows users to create other ConfigMaps for configurations such as plugins or storage